iFixit/react-commerce
# Migrate product pages to app router

This PR migrates product pages from `pages` to `app` folder, marking a substantial shift in our structure as it introduces React Server Components (RSC). This is more than just a file relocation—it opens up a different caching strategy, a distinct routing approach, a possible large reduction in client bundle size and several other core changes.

In order to ensure a seamless transition, the changes in this PR were kept lean and minimalistic.
Future optimizations are in sight, and they involve transforming a significant segment of the UI into RSCs to prevent unnecessary code from permeating into the client bundle. This however, lies beyond the scope of this specific PR and will be addressed subsequently.

Most of the changes you will see in this PR are linked to these 4 requirements

- Data Retrieval: We had to transition from leveraging the getServerSideProps function to using a plain JavaScript function invoked within a React Server Component (RSC).
- SEO and Metadata: Since the `<Head/>` component is no longer available, metadata is now populated using the [metadata API](https://nextjs.org/docs/app/building-your-application/optimizing/metadata).
- Routing Changes: The new router has a different API. For example the request query and headers are not returned by the `useRouter` hook by default. Those data are now readable only via specific functions (`headers()` and `useSearchParams()`). 
- Caching Headers: given that headers retrieved via `headers()` function are now read-only, we are precluded from setting caching headers as we previously did with getServerSideProps.

A couple of things to monitor:

- Shallow Navigation: As highlighted in this [github discussion](https://github.com/vercel/next.js/discussions/48110#discussioncomment-6481618), shallow navigation is not yet available. Thus, we experience a slight UI delay due to re-rendering whenever URL parameters get altered via the new router.
- Data Cache: As mentioned in the documentation, a fetch call that follows a headers call will disable data caching. Therefore, should we decide to use data caching in the future, we must consider this aspect.

### QA

1. Visit Vercel preview
2. Check the new product page to highlight potential differences with the previous one.
3. If you try to make a performance comparison, consider that preview branches might include additional js due to the Vercel UI for commenting previews.
